shotover/src/server.rs

```diff
@@ -276,7 +294,64 @@ impl<C: CodecBuilder + 'static> TcpCodecListener<C> {
     }
 
     pub async fn shutdown(&mut self) {
-        join_all(&mut self.connection_handles).await;
+        // First, signal all connections to shutdown
+        for tc in &self.connection_handles {
+            tc.shutdown();
+        }
+
+        // Then wait for all connections to complete
+        join_all(self.connection_handles.iter_mut().map(|tc| &mut tc.handle)).await;
+    }
+
+    pub async fn gradual_shutdown(&mut self) {
+        info!(
+            "[{}] Gradual shutdown initiated - draining connections",
+            self.source_name
+        );
+
+        // Calculate 10% of connections to drain (at least 1 if there are any connections)
+        let connections_to_drain = std::cmp::max(
+            1,
+            (self.connection_handles.len() as f64 * 0.1).ceil() as usize,
+        );
+
+        while !self.connection_handles.is_empty() {
+            let to_drain = std::cmp::min(connections_to_drain, self.connection_handles.len());
+
+            info!(
+                "[{}] Draining {} out of {} connections",
+                self.source_name,
+                to_drain,
+                self.connection_handles.len()
+            );
+
+            // Drain the first N connections and shut them down
+            let mut connections_to_close: Vec<_> =
+                self.connection_handles.drain(..to_drain).collect();
+            for connection in &connections_to_close {
+                connection.shutdown();
+            }
+
+            // Wait for all shutdown connections to complete
+            join_all(connections_to_close.iter_mut().map(|tc| &mut tc.handle)).await;
+
+            info!(
+                "[{}] {} connections remaining after drain",
+                self.source_name,
+                self.connection_handles.len()
+            );
+
+            // Don't sleep after draining the last batch
+            if !self.connection_handles.is_empty() {
```

Comment:
good thinking!

I also appreciate fixing the `broadcast::Receiver` ->  `watch::Receiver` below.